File: android/src/main/java/com/github/marekchen/flutterqq/FlutterQqPlugin.java
package com.github.marekchen.flutterqq;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.tencent.tauth.IUiListener;
import com.tencent.tauth.Tencent;
import com.tencent.tauth.UiError;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.ActivityResultListener;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import org.json.JSONObject;


public final class FlutterQqPlugin implements FlutterPlugin, ActivityAware, MethodCallHandler {
    private MethodChannel mMethodChannel;
    private ActivityPluginBinding mActivityBinder;
    private boolean isLogin;
    private static Tencent mTencent;

    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        this.mMethodChannel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "flutter_qq");
        MethodChannel var10000 = this.mMethodChannel;
        if (var10000 != null) {
            var10000.setMethodCallHandler((MethodCallHandler)this);
        }

    }

    public void onDetachedFromEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        MethodChannel var10000 = this.mMethodChannel;
        if (var10000 != null) {
            var10000.setMethodCallHandler((MethodCallHandler)null);
        }

        this.mMethodChannel = (MethodChannel)null;
    }

    public void onAttachedToActivity(@NonNull ActivityPluginBinding activityPluginBinding) {
        this.mActivityBinder = activityPluginBinding;
    }

    public void onDetachedFromActivityForConfigChanges() {
        this.mActivityBinder = (ActivityPluginBinding)null;
    }

    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding activityPluginBinding) {
        this.onAttachedToActivity(activityPluginBinding);
    }

    public void onDetachedFromActivity() {
        this.onDetachedFromActivityForConfigChanges();
    }

    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        FlutterQqPlugin.OneListener listener = new FlutterQqPlugin.OneListener();
        ActivityPluginBinding var10000 = this.mActivityBinder;
        if (var10000 != null) {
            var10000.addActivityResultListener((ActivityResultListener)listener);
        }

        String var5 = call.method;
        if (var5 != null) {
            String var4 = var5;
            switch(var4.hashCode()) {
                case -1869931517:
                    if (var4.equals("registerQQ")) {
                        this.registerQQ(call, result);
                    }
                    break;
                case -1585587965:
                    if (var4.equals("shareToQzone")) {
                        this.isLogin = false;
                        listener.setResult$flutter_qq_debug(result);
                        this.doShareToQzone(call, listener);
                    }
                    break;
                case -1582030246:
                    if (var4.equals("shareToQQ")) {
                        this.isLogin = false;
                        listener.setResult$flutter_qq_debug(result);
                        this.doShareToQQ(call, listener);
                    }
                    break;
                case 103149417:
                    if (var4.equals("login")) {
                        this.isLogin = true;
                        listener.setResult$flutter_qq_debug(result);
                        this.login(call, listener);
                    }
                    break;
                case 693216720:
                    if (var4.equals("isQQInstalled")) {
                        this.isQQInstalled(call, result);
                    }
            }
        }

    }

    private final void registerQQ(MethodCall call, Result result) {
        String mAppid = (String)call.argument("appId");
        ActivityPluginBinding var10001 = this.mActivityBinder;
        if (var10001 != null) {
            Activity var4 = var10001.getActivity();
            if (var4 != null) {
                mTencent = Tencent.createInstance(mAppid, (Context)var4);
                result.success(true);
                return;
            }
        }

    }

    private final void isQQInstalled(MethodCall call, Result result) {
        Tencent var10001 = mTencent;
        Boolean var3;
        if (var10001 != null) {
            ActivityPluginBinding var10002 = this.mActivityBinder;
            if (var10002 == null) {
                return;
            }

            Activity var4 = var10002.getActivity();
            if (var4 == null) {
                return;
            }

            var3 = var10001.isQQInstalled((Context)var4);
        } else {
            var3 = null;
        }

        result.success(var3);
    }

    private final void login(MethodCall call, FlutterQqPlugin.OneListener listener) {
        String scopes = (String)call.argument("scopes");
        Tencent var10000 = mTencent;
        if (var10000 != null) {
            ActivityPluginBinding var10001 = this.mActivityBinder;
            if (var10001 == null) {
                return;
            }

            Activity var4 = var10001.getActivity();
            if (var4 == null) {
                return;
            }

            String var10002 = scopes;
            if (scopes == null) {
                var10002 = "get_simple_userinfo";
            }

            var10000.login(var4, var10002, (IUiListener)listener);
        }

    }

    private final void doShareToQQ(MethodCall call, final FlutterQqPlugin.OneListener listener) {
        final Bundle params;
        Integer shareType;
        byte var5;
        label39: {
            params = new Bundle();
            shareType = (Integer)call.argument("shareType");
            Log.i("FlutterQqPlugin", "arguments:" + call.arguments);
            var5 = 5;
            if (shareType != null) {
                if (shareType == var5) {
                    break label39;
                }
            }

            params.putString("title", (String)call.argument("title"));
            params.putString("targetUrl", (String)call.argument("targetUrl"));
            params.putString("summary", (String)call.argument("summary"));
        }

        label34: {
            var5 = 5;
            if (shareType != null) {
                if (shareType == var5) {
                    params.putString("imageLocalUrl", (String)call.argument("imageLocalUrl"));
                    break label34;
                }
            }

            params.putString("imageUrl", (String)call.argument("imageUrl"));
        }

        params.putString("appName", (String)call.argument("appName"));
        params.putInt("req_type", shareType != null ? shareType : 0);
        Integer var10002 = (Integer)call.argument("qzoneFlag");
        if (var10002 == null) {
            var10002 = 0;
        }

        params.putInt("cflag", var10002);
        var5 = 2;
        if (shareType != null) {
            if (shareType == var5) {
                params.putString("audio_url", (String)call.argument("audioUrl"));
            }
        }

        params.putString("share_to_qq_ark_info", (String)call.argument("ark"));
        Log.i("FlutterQqPlugin", "params:" + params);
        (new Handler(Looper.getMainLooper())).post((Runnable)(new Runnable() {
            public final void run() {
                Tencent var10000 = FlutterQqPlugin.mTencent;
                if (var10000 != null) {
                    ActivityPluginBinding var10001 = FlutterQqPlugin.this.mActivityBinder;
                    if (var10001 == null) {
                        return;
                    }

                    Activity var1 = var10001.getActivity();
                    if (var1 == null) {
                        return;
                    }

                    var10000.shareToQQ(var1, params, (IUiListener)listener);
                }

            }
        }));
    }

    private final void doShareToQzone(MethodCall call, final FlutterQqPlugin.OneListener listener) {
        final Bundle params = new Bundle();
        Integer shareType = (Integer)call.argument("shareType");
        Log.i("FlutterQqPlugin", "arguments:" + call.arguments);
        params.putInt("req_type", shareType != null ? shareType : 0);
        params.putString("title", (String)call.argument("title"));
        params.putString("summary", (String)call.argument("summary"));
        params.putString("targetUrl", (String)call.argument("targetUrl"));
        ArrayList list = new ArrayList();
        String var10001 = (String)call.argument("imageUrl");
        if (var10001 != null) {
            list.add(var10001);
            params.putStringArrayList("imageUrl", list);
            params.putString("videoPath", (String)call.argument("videoPath"));
            Bundle bundle2 = new Bundle();
            bundle2.putString("hulian_extra_scene", (String)call.argument("scene"));
            bundle2.putString("hulian_call_back", (String)call.argument("hulian_call_back"));
            params.putBundle("extMap", bundle2);
            Log.i("FlutterQqPlugin", "params:" + params);
            byte var7 = 1;
            if (shareType != null) {
                if (shareType == var7) {
                    (new Handler(Looper.getMainLooper())).post((Runnable)(new Runnable() {
                        public final void run() {
                            Tencent var10000 = FlutterQqPlugin.mTencent;
                            if (var10000 != null) {
                                ActivityPluginBinding var10001 = FlutterQqPlugin.this.mActivityBinder;
                                if (var10001 == null) {
                                    return;
                                }

                                Activity var1 = var10001.getActivity();
                                if (var1 == null) {
                                    return;
                                }

                                var10000.shareToQzone(var1, params, (IUiListener)listener);
                            }

                        }
                    }));
                    return;
                }
            }

            (new Handler(Looper.getMainLooper())).post((Runnable)(new Runnable() {
                public final void run() {
                    Tencent var10000 = FlutterQqPlugin.mTencent;
                    if (var10000 != null) {
                        ActivityPluginBinding var10001 = FlutterQqPlugin.this.mActivityBinder;
                        if (var10001 == null) {
                            return;
                        }

                        Activity var1 = var10001.getActivity();
                        if (var1 == null) {
                            return;
                        }

                        var10000.publishToQzone(var1, params, (IUiListener)listener);
                    }

                }
            }));
        }
    }

    // $FF: synthetic method
    public static final void access$setLogin$p(FlutterQqPlugin $this, boolean var1) {
        $this.isLogin = var1;
    }

    // $FF: synthetic method
    public static final void access$setMTencent$cp(Tencent var0) {
        mTencent = var0;
    }

    // $FF: synthetic method
    public static final void access$setMActivityBinder$p(FlutterQqPlugin $this, ActivityPluginBinding var1) {
        $this.mActivityBinder = var1;
    }

    private final class OneListener implements IUiListener, ActivityResultListener {
        private Result result;

        public final void setResult$flutter_qq_debug(@NonNull Result result) {
            this.result = result;
        }

        public void onComplete(@Nullable Object response) {
            Log.i("FlutterQqPlugin", String.valueOf(response));
            HashMap re = new HashMap();
            Result var7;
            if (!FlutterQqPlugin.this.isLogin) {
                ((Map)re).put("Code", 0);
                ((Map)re).put("Message", String.valueOf(response));
                var7 = this.result;
                if (var7 != null) {
                    var7.success(re);
                }

            } else if (response == null) {
                ((Map)re).put("Code", 1);
                ((Map)re).put("Message", "response is empty");
                var7 = this.result;
                if (var7 != null) {
                    var7.success(re);
                }

            } else {
                Object var10000 = response;
                if (!(response instanceof JSONObject)) {
                    var10000 = null;
                }

                JSONObject jsonResponse = (JSONObject)var10000;
                if (jsonResponse != null && jsonResponse.length() != 0) {
                    HashMap resp = new HashMap();

                    String var10002;
                    Map var8;
                    try {
                        Log.i("FlutterQqPlugin", resp.toString());
                        var8 = (Map)resp;
                        var10002 = jsonResponse.getString("openid");
                        var8.put("openid", var10002);
                        var8 = (Map)resp;
                        var10002 = jsonResponse.getString("access_token");
                        var8.put("accessToken", var10002);
                        ((Map)resp).put("expiresAt", jsonResponse.getLong("expires_time"));
                        ((Map)re).put("Code", 0);
                        ((Map)re).put("Message", "ok");
                        ((Map)re).put("Response", resp);
                        var7 = this.result;
                        if (var7 != null) {
                            var7.success(re);
                        }

                    } catch (Exception var6) {
                        ((Map)re).put("Code", 1);
                        var8 = (Map)re;
                        var10002 = var6.getLocalizedMessage();
                        var8.put("Message", var10002);
                        var7 = this.result;
                        if (var7 != null) {
                            var7.success(re);
                        }

                    }
                } else {
                    ((Map)re).put("Code", 1);
                    ((Map)re).put("Message", "response is empty");
                    var7 = this.result;
                    if (var7 != null) {
                        var7.success(re);
                    }

                }
            }
        }

        public void onError(@Nullable UiError uiError) {
            Log.w("FlutterQqPlugin", "errorCode:" + (uiError != null ? uiError.errorCode : null) + ";errorMessage:" + (uiError != null ? uiError.errorMessage : null));
            HashMap re = new HashMap();
            ((Map)re).put("Code", 1);
            ((Map)re).put("Message", "errorCode:" + (uiError != null ? uiError.errorCode : null) + ";errorMessage:" + (uiError != null ? uiError.errorMessage : null));
            Result var10000 = this.result;
            if (var10000 != null) {
                var10000.success(re);
            }

        }

        public void onCancel() {
            Log.w("FlutterQqPlugin", "error:cancel");
            HashMap re = new HashMap();
            ((Map)re).put("Code", 2);
            ((Map)re).put("Message", "cancel");
            Result var10000 = this.result;
            if (var10000 != null) {
                var10000.success(re);
            }

        }

        public boolean onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
            if (requestCode != 11101 && requestCode != 10103 && requestCode != 10104 && requestCode != 10102) {
                return false;
            } else {
                Tencent.onActivityResultData(requestCode, resultCode, data, (IUiListener)this);
                return true;
            }
        }

        public OneListener() {
        }
    }


}

File: android/src/main/java/com/github/marekchen/flutterqq/FlutterQqPlugin.kt
package com.github.marekchen.flutterqq

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log

import com.tencent.connect.common.Constants
import com.tencent.connect.share.QQShare
import com.tencent.connect.share.QzonePublish
import com.tencent.connect.share.QzoneShare
import com.tencent.tauth.IUiListener
import com.tencent.tauth.Tencent
import com.tencent.tauth.UiError

import org.json.JSONObject

import java.util.ArrayList
import java.util.HashMap

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.PluginRegistry

/**
 * FlutterQqPlugin
 */
class FlutterQqPlugin : FlutterPlugin, ActivityAware, MethodCallHandler {

    private var mMethodChannel: MethodChannel? = null
    private var mActivityBinder: ActivityPluginBinding? = null
    private var isLogin: Boolean = false

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        mMethodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_qq")
        mMethodChannel?.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        mMethodChannel?.setMethodCallHandler(null)
        mMethodChannel = null
    }

    override fun onAttachedToActivity(activityPluginBinding: ActivityPluginBinding) {
        mActivityBinder = activityPluginBinding
    }

    override fun onDetachedFromActivityForConfigChanges() {
        mActivityBinder = null
    }

    override fun onReattachedToActivityForConfigChanges(activityPluginBinding: ActivityPluginBinding) {
        onAttachedToActivity(activityPluginBinding)
    }

    override fun onDetachedFromActivity() {
        onDetachedFromActivityForConfigChanges()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        val listener = OneListener()
        mActivityBinder?.addActivityResultListener(listener)
        when (call.method) {
            "registerQQ" -> registerQQ(call, result)
            "isQQInstalled" -> isQQInstalled(call, result)
            "login" -> {
                isLogin = true
                listener.setResult(result)
                login(call, listener)
            }
            "shareToQQ" -> {
                isLogin = false
                listener.setResult(result)
                doShareToQQ(call, listener)
            }
            "shareToQzone" -> {
                isLogin = false
                listener.setResult(result)
                doShareToQzone(call, listener)
            }
        }
    }

    private fun registerQQ(call: MethodCall, result: Result) {
        val mAppid = call.argument<String>("appId")
        mTencent = Tencent.createInstance(mAppid, mActivityBinder?.activity?:return)
        result.success(true)
    }

    private fun isQQInstalled(call: MethodCall, result: Result) {
        result.success(mTencent?.isQQInstalled(mActivityBinder?.activity?:return))
    }

    private fun login(call: MethodCall, listener: OneListener) {
        val scopes = call.argument<Any>("scopes") as String?
        mTencent?.login(mActivityBinder?.activity?:return, scopes ?: "get_simple_userinfo", listener)
    }

    private fun doShareToQQ(call: MethodCall, listener: OneListener) {
        val params = Bundle()
        val shareType = call.argument<Int>("shareType")
        Log.i("FlutterQqPlugin", "arguments:" + call.arguments)
        if (shareType != QQShare.SHARE_TO_QQ_TYPE_IMAGE) {
            params.putString(QQShare.SHARE_TO_QQ_TITLE, call.argument<String>("title"))
            params.putString(QQShare.SHARE_TO_QQ_TARGET_URL, call.argument<String>("targetUrl"))
            params.putString(QQShare.SHARE_TO_QQ_SUMMARY, call.argument<String>("summary"))
        }
        if (shareType == QQShare.SHARE_TO_QQ_TYPE_IMAGE) {
            params.putString(QQShare.SHARE_TO_QQ_IMAGE_LOCAL_URL, call.argument<String>("imageLocalUrl"))
        } else {
            params.putString(QQShare.SHARE_TO_QQ_IMAGE_URL, call.argument<String>("imageUrl"))
        }
        params.putString(QQShare.SHARE_TO_QQ_APP_NAME, call.argument<String>("appName"))
        params.putInt(QQShare.SHARE_TO_QQ_KEY_TYPE, shareType?:0)
        params.putInt(QQShare.SHARE_TO_QQ_EXT_INT, (call.argument<Int>("qzoneFlag")?:0))
        if (shareType == QQShare.SHARE_TO_QQ_TYPE_AUDIO) {
            params.putString(QQShare.SHARE_TO_QQ_AUDIO_URL, call.argument<String>("audioUrl"))
        }
        params.putString(QQShare.SHARE_TO_QQ_ARK_INFO, call.argument<String>("ark"))
        Log.i("FlutterQqPlugin", "params:$params")
        Handler(Looper.getMainLooper()).post { mTencent?.shareToQQ(mActivityBinder?.activity?:return@post, params, listener) }
    }

    private fun doShareToQzone(call: MethodCall, listener: OneListener) {
        val params = Bundle()
        val shareType = call.argument<Int>("shareType")
        Log.i("FlutterQqPlugin", "arguments:" + call.arguments)
        params.putInt(QzoneShare.SHARE_TO_QZONE_KEY_TYPE, shareType?:0)
        params.putString(QzoneShare.SHARE_TO_QQ_TITLE, call.argument<String>("title"))
        params.putString(QzoneShare.SHARE_TO_QQ_SUMMARY, call.argument<String>("summary") )
        params.putString(QzoneShare.SHARE_TO_QQ_TARGET_URL, call.argument<String>("targetUrl"))
        val list = ArrayList<String>()
        list.add(call.argument("imageUrl")?:return)
        params.putStringArrayList(QQShare.SHARE_TO_QQ_IMAGE_URL, list)
        //params.putStringArrayList(QQShare.SHARE_TO_QQ_IMAGE_URL, (ArrayList<String>) call.argument("imageUrls"));
        params.putString(QzonePublish.PUBLISH_TO_QZONE_VIDEO_PATH, call.argument<String>("videoPath"))
        val bundle2 = Bundle()
        bundle2.putString(QzonePublish.HULIAN_EXTRA_SCENE, call.argument<String>("scene"))
        bundle2.putString(QzonePublish.HULIAN_CALL_BACK, call.argument<String>("hulian_call_back"))
        params.putBundle(QzonePublish.PUBLISH_TO_QZONE_EXTMAP, bundle2)
        Log.i("FlutterQqPlugin", "params:$params")
        if (shareType == QzoneShare.SHARE_TO_QZONE_TYPE_IMAGE_TEXT) {
            Handler(Looper.getMainLooper()).post { mTencent?.shareToQzone(mActivityBinder?.activity?:return@post, params, listener) }
        } else {
            Handler(Looper.getMainLooper()).post { mTencent?.publishToQzone(mActivityBinder?.activity?:return@post, params, listener) }
        }
    }

    private inner class OneListener : IUiListener, PluginRegistry.ActivityResultListener {

        private var result: Result? = null

        internal fun setResult(result: Result) {
            this.result = result
        }

        override fun onComplete(response: Any?) {
            Log.i("FlutterQqPlugin", response.toString())
            val re = HashMap<String, Any>()
            if (isLogin) {
                if (null == response) {
                    re["Code"] = 1
                    re["Message"] = "response is empty"
                    result?.success(re)
                    return
                }
                val jsonResponse = response as? JSONObject
                if (null == jsonResponse || jsonResponse.length() == 0) {
                    re["Code"] = 1
                    re["Message"] = "response is empty"
                    result?.success(re)
                    return
                }
                val resp = HashMap<String, Any>()
                try {
                    Log.i("FlutterQqPlugin", resp.toString())
                    resp["openid"] = jsonResponse.getString(Constants.PARAM_OPEN_ID)
                    resp["accessToken"] = jsonResponse.getString(Constants.PARAM_ACCESS_TOKEN)
                    resp["expiresAt"] = jsonResponse.getLong(Constants.PARAM_EXPIRES_TIME)
                    // resp.put("appId", jsonResponse.getString(Constants.PARAM_APP_ID));
                    re["Code"] = 0
                    re["Message"] = "ok"
                    re["Response"] = resp
                    result?.success(re)
                    return
                } catch (e: Exception) {
                    re["Code"] = 1
                    re["Message"] = e.localizedMessage
                    result?.success(re)
                    return
                }

            }
            re["Code"] = 0
            re["Message"] = response.toString()
            result?.success(re)
        }

        override fun onError(uiError: UiError?) {
            Log.w("FlutterQqPlugin", "errorCode:${uiError?.errorCode};errorMessage:${uiError?.errorMessage}")
            val re = HashMap<String, Any>()
            re["Code"] = 1
            re["Message"] = "errorCode:${uiError?.errorCode};errorMessage:${uiError?.errorMessage}"
            result?.success(re)
        }

        override fun onCancel() {
            Log.w("FlutterQqPlugin", "error:cancel")
            val re = HashMap<String, Any>()
            re["Code"] = 2
            re["Message"] = "cancel"
            result?.success(re)
        }

        override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
            if (requestCode == Constants.REQUEST_LOGIN ||
                    requestCode == Constants.REQUEST_QQ_SHARE ||
                    requestCode == Constants.REQUEST_QZONE_SHARE ||
                    requestCode == Constants.REQUEST_APPBAR) {
                Tencent.onActivityResultData(requestCode, resultCode, data, this)
                return true
            }
            return false
        }

    }

    companion object {
        private var mTencent: Tencent? = null
    }


}
